trunk/tennix.6
.TH "TENNIX" "6" "February 2009" "tennix 0.7.0"
 
.SH "NAME"
tennix \(em A funny 2D tennis game

.SH "SYNOPSIS"
.B tennix
.RI [ options ]

.SH "DESCRIPTION"
.PP 
\fBtennix\fR is a top-view 2D tennis game. It  
features a two-player game mode and a single-player mode against the 
computer.  
.PP 
Joystick can be used instead of the keyboard. See \fBOPTIONS\fR on how to
set up the game for joysticks.
.PP 
This is the SDL port of a DOS game written in 2003. The original game can 
still be found on the web at http://thpinfo.com/2003/tennix/

.SH "CONTROLS"
.TP
.I W, S, D
Player 1: Move up, down and swing the racket
.TP
.I Up, Down, Spacebar
Player 1 (alternative): Move up, down and swing the racket
.TP
.I Mouse, Left mouse button
Player 1 (alternative): Move while left button pressed, swing racket when left button released
.TP
.I O, L, K
Player 2: Move up, down and swing the racket
.TP
.I C
Change the type of court (grass, hard, clay)
.TP
.I P
(Un)pause the game
.TP
.I F
Switch between windowed and fullscreen mode
.SH "OPTIONS"
.TP
.I \-f, \-\-fullscreen
Start the game in fullscreen mode
.TP
.I \-h, \-\-help
Show summary of options
.TP
.I \-J, \-\-list-joysticks
Shows a list of joysticks and how to use them
.TP
.I \-j,  \-\-joystick \fBName\fP
Choose the joystick to be used to control Player 1
.SH "HOMEPAGE" 
.PP 
http://icculus.org/tennix/ 
.SH "AUTHOR" 
.PP 
The author and current maintainer of Tennix is Thomas Perl (thp@thpinfo.com).
.PP 
This manual page was written by Andrea Colangelo (warp10@ubuntu.com) for 
the \fBUbuntu\fP system (but may be used by others).  Permission is 
granted to copy, distribute and/or modify this document under 
the terms of the GNU General Public License, Version 2 any  
later version published by the Free Software Foundation. 
.PP 
On Debian systems, the complete text of the GNU General Public 
License can be found in /usr/share/common-licenses/GPL-2. 



ChangeLog
commit 749597197a66e94f8742b9ee90dfabafd59d4dc3
Author: Thomas Perl <thp@perli.net>
Date:   Fri Feb 6 00:34:21 2009 +0100

    Tennix 0.7.0 "Deine Eltern sind auf einem Tennixturnier"
    
    Updated README and manpage. Made sure tennix.tnx is
    installed in the right spot, and that the binary can
    find the data file either in the current directory or
    in the installed location.
    
    Push version to 0.7.0 and display version and URL in
    the main screen (as with previous versions).

commit a6ae7d2deba542bba9d3b8379b237323e33f679b
Author: Thomas Perl <thp@perli.net>
Date:   Sat Jan 31 13:55:57 2009 +0100

    Updated year, README, TODO list
    
    Added 2009 to copyright info where necessary
    Updated README file
    Updated TODO list

commit 9d28bad7001375e56b8c10180ba5bbc3d3eba862
Author: Thomas Perl <thp@perli.net>
Date:   Sat Jan 31 13:44:56 2009 +0100

    Documentation: How to generate callgraphs for Tennix
    
    Remove the -pg option and make the callgraph-generating
    code optional in the makefile. See doc/callgraph.txt for
    more information.

commit c1ee5818ba1684854a8a40b1585406ff309c4ee7
Author: Thomas Perl <thp@perli.net>
Date:   Tue Jan 27 11:10:40 2009 +0100

    Add makefile option for update rectangles mode
    
    This mode visualizes which screen regions are updated
    on each frame. Compile with: make UPDRECTANGLE=1

commit 42e6c49f66e7a81984147af75879158184361fd4
Author: Thomas Perl <thp@perli.net>
Date:   Tue Jan 27 10:58:18 2009 +0100

    Add compiler flags for Maemo

commit 3c75fdba48e9ac4f5873de10831d96f88bbdecac
Author: Thomas Perl <thp@perli.net>
Date:   Tue Jan 27 00:08:50 2009 +0100

    Only redraw the referee when its state changes
    
    Same principle as with the score string - remember the
    last "position" of the referee and if it has changed,
    redraw the back buffer. This saves a lot of painting.

commit 59369f27cc8c14e7a191a56e71c2dc015d61ab77
Author: Thomas Perl <thp@perli.net>
Date:   Tue Jan 27 00:04:45 2009 +0100

    Only redraw game score strings when necessary
    
    Do not redraw the score strings every time the
    render() method is called, but only when the text
    really changes (using a flag in the GameState).

commit 312dfc15bbc58a058d97468a86af938c8e28f98b
Author: Thomas Perl <thp@perli.net>
Date:   Mon Jan 26 23:41:23 2009 +0100

    Remove is_fading() from the graphics module
    
    Only used in two places, now refactored.

commit 91414e23c4e9aeef14d6bde1c44de9a7bc07a531
Author: Thomas Perl <thp@perli.net>
Date:   Wed Jan 21 23:19:59 2009 +0100

    Updated pygame test code

commit 4eb9fc349b210bcdfa96dd2c91d6e8b4f1424cd4
Author: Thomas Perl <thp@perli.net>
Date:   Mon Jan 19 19:33:20 2009 +0100

    Add "--benchmark/-b" command line switch
    
    When starting tennix with "--benchmark" or "-b", the
    game will start in "AI vs. AI" mode for a limited
    number of seconds. This can be used for benchmarking
    different optimizations.

commit 78baa1fe95365ec91b06563fc42d0b670a55909c
Author: Thomas Perl <thp@perli.net>
Date:   Thu Jan 15 12:10:45 2009 +0100

    Make sure graphics and sound data is really freed
    
    After loading of graphics and sound, make sure to
    unload the allocated resources, because RWops does
    not seem to free the resources, even if told so.

commit 3c90f67e8b80237ecd8b1998b5c9b4d8c5cd2273
Author: Thomas Perl <thp@perli.net>
Date:   Fri Jan 9 15:01:13 2009 +0100

    Better loading progress bar (uses button drawing)

commit 7bcd5b3dd3ab24e2e09bfe78582a195c809af846
Author: Thomas Perl <thp@perli.net>
Date:   Fri Jan 9 15:00:40 2009 +0100

    Fix referee talk image usage; support more ball states

commit 85f87a017531a47e5b2389c1b3527b4522a29986
Author: Thomas Perl <thp@perli.net>
Date:   Thu Jan 8 19:44:57 2009 +0100

    Loading progress bar, add missing header files, Maemo
    
    Add a progress bar while loading resources (graphics
    and sound) from the datafile.
    
    Add some missing "#include" lines (scratchbox complained).
    
    Do not make the "hovering" sound on Maemo (touchscreen!)

commit c734b6b9be6d860c3cabd1d513e2980da4f33a69
Author: Thomas Perl <thp@perli.net>
Date:   Thu Jan 8 19:32:47 2009 +0100

    Cache width and xposition in get_font_metrics
    
    To save some cycles, do not re-calculate the font
    metrics every time a string is drawn, but save the
    width and xposition permanently after the first
    calculation to speed up font drawing.

commit 22eb5ce6d33ee8cc392aa209b3eb0baf1654de8d
Author: Thomas Perl <thp@perli.net>
Date:   Wed Jan 7 15:12:43 2009 +0100

    Use Tennix Archive Format instead in-executable data
    
    This makes the binary smaller, and we can free the
    raw game data once the resources have been loaded.
    
    Also, it makes recompiling easier. The data file is
    currently named "tennix.tnx" and will be auto-built.

commit abd420bce6014073ca49aebcc50127fed09c9ece
Author: Thomas Perl <thp@perli.net>
Date:   Wed Jan 7 14:09:38 2009 +0100

    Voice support, experimental wind, extended referee
    
    Add support for playing back voice tracks if voice
    files are available (have to enable with "DELUXE=1").
    
    Includes experimental support for in-game wind (you
    can modify its direction with "1" and "2").
    
    Add support for the extended referee mode if the
    extended referee resources are available (DELUXE=1).

commit 54f9b4f98982a83c14c967886b78abf262c09d43
Author: Thomas Perl <thp@perli.net>

trunk/tennix.c

/**
 *
 * Tennix! SDL Port
 * Copyright (C) 2003, 2007, 2008, 2009 Thomas Perl <thp@thpinfo.com>
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, 
 * MA  02110-1301, USA.
 *
 **/

#include <stdio.h>
#include <time.h>
#include <libgen.h>
#include <string.h>
#include <stdlib.h>

#ifdef WIN32
#include <windows.h>
#endif

#include "tennix.h"
#include "game.h"
#include "graphics.h"
#include "sound.h"
#include "input.h"

SDL_Surface *screen;

#ifdef WIN32

/* IDs from the resource file */
#define START_BUTTON 1
#define CHECKBOX_FULLSCREEN 2
#define QUIT_BUTTON 3

BOOL CALLBACK ConfigDialogProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
{
    static int checkbox_is_checked;

    switch (uMsg) {
        case WM_CLOSE:
            EndDialog(hwndDlg, IDCANCEL);
            break;
        case WM_COMMAND:
            switch (wParam) {
                case START_BUTTON:
                    EndDialog(hwndDlg, (checkbox_is_checked)?(IDYES):(IDNO));
                    break;
                case QUIT_BUTTON:
                    EndDialog(hwndDlg, IDCANCEL);
                    break;
                case CHECKBOX_FULLSCREEN:
                    checkbox_is_checked ^= 1;
                    break;
            }
            break;
        default:
            return FALSE;
    }
    return TRUE;
}
#endif

#ifdef WIN32
int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, 
    LPSTR lpCmdLine, int nCmdShow) {
#else
int main( int argc, char** argv) {
#endif
    int i, slide, slide_direction;
    int ticks;
    int mx, my;
    Uint8 *keys;
    Uint8 mb;
    SDL_Event e;
    int sdl_flags = SDL_SWSURFACE;
    int btn_hovering = 0, btn_hovering_old = 0;
    int slide_start;
    bool mouse_pressed = false;
    bool quit = false;
    bool benchmark = false;
    GameState *current_game = NULL, *prepared_game = NULL;

    MenuButton btn_back = {
        "Back to main menu",
        MENU_OPTIONS_BORDER,
        HEIGHT-(MENU_OPTIONS_BORDER+MENU_OPTIONS_BUTTON_HEIGHT),
        MENU_OPTIONS_BUTTON_WIDTH, MENU_OPTIONS_BUTTON_HEIGHT,
        255, 0, 0
    };
    MenuButton btn_start = {
        "Start new game",
        WIDTH-(MENU_OPTIONS_BORDER+MENU_OPTIONS_BUTTON_WIDTH),
        HEIGHT-(MENU_OPTIONS_BORDER+MENU_OPTIONS_BUTTON_HEIGHT),
        MENU_OPTIONS_BUTTON_WIDTH, MENU_OPTIONS_BUTTON_HEIGHT,
        0, 255, 0
    };
    MenuButton btn_court_change = {
        "change",
        240,
        40,
        50, MENU_OPTIONS_BUTTON_HEIGHT,
        100, 100, 100
    };
    MenuButton btn_player1 = {
        NULL,
        380,
        180,
        MENU_OPTIONS_BUTTON_WIDTH, MENU_OPTIONS_BUTTON_HEIGHT,
        50, 50, 255
    };
    MenuButton btn_player2 = {
        NULL,
        380,
        180+MENU_OPTIONS_BORDER*2+MENU_OPTIONS_BUTTON_HEIGHT,
        MENU_OPTIONS_BUTTON_WIDTH, MENU_OPTIONS_BUTTON_HEIGHT,
        255, 50, 50
    };

    int state = MENU_STATE_STARTED;

#ifdef ENABLE_FPS_LIMIT
    Uint32 ft, frames; /* frame timer and frames */
#endif

#ifdef MAEMO
    sdl_flags |= SDL_FULLSCREEN;
#endif

#ifdef WIN32
    int mb_result;
    mb_result = DialogBox(hInstance, "CONFIG", 0, (DLGPROC)ConfigDialogProc);

    switch (mb_result) {
        case IDYES:
            sdl_flags |= SDL_FULLSCREEN;
            break;
        case IDCANCEL:
            return 0;
            break;
        default:
            break;
    }
#else
    fprintf(stderr, "Tennix " VERSION "\n" COPYRIGHT "\n" URL "\n\n");

    bool do_help = false;
    i = 1;
    while (i < argc) {
        /* A poor/lazy man's getopt */
        #define OPTION_SET(longopt,shortopt) \
                (strcmp(argv[i], longopt)==0 || strcmp(argv[i], shortopt)==0)
        #define OPTION_VALUE \
                ((i+1 < argc)?(argv[i+1]):(NULL))
        #define OPTION_VALUE_PROCESSED \
                (i++)
        if (OPTION_SET("--fullscreen", "-f")) {
            sdl_flags |= SDL_FULLSCREEN;
        }
        else if (OPTION_SET("--help", "-h")) {
            do_help = true;
        }
        else if (OPTION_SET("--list-joysticks", "-J")) {
            SDL_Init(SDL_INIT_JOYSTICK);
            joystick_list();
            return 0;
        }
        else if (OPTION_SET("--benchmark", "-b")) {
            benchmark = true;
        }
        else if (OPTION_SET("--joystick", "-j")) {
            SDL_Init(SDL_INIT_JOYSTICK);
            if (OPTION_VALUE == NULL) {
                fprintf(stderr, "Error: You need to specify the name of the joystick as parameter.\n");
                do_help = true;
                break;
            }
            if (joystick_open(OPTION_VALUE)==0) {
                fprintf(stderr, "Warning: Cannot find joystick \"%s\" - Ignored.\n", OPTION_VALUE);
                break;
            }
            OPTION_VALUE_PROCESSED;
        }
        else {
            fprintf(stderr, "Ignoring unknown option: %s\n", argv[i]);
        }
        i++;
    }

    if (do_help == true) {
        fprintf(stderr, "Usage: %s [--fullscreen|-f] [--help|-h] [--list-joysticks|-J] [--joystick|-j] [joystick name]\n", argv[0]);
        return 0;
    }
#endif

    if (benchmark) {
        srand(100);
    } else {
        srand((unsigned)time(NULL));
    }

    if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_JOYSTICK) == -1) {
        fprintf( stderr, "Can't init SDL:  %s\n", SDL_GetError());
        exit( 1);
    }

    SDL_VideoInfo* vi = (SDL_VideoInfo*)SDL_GetVideoInfo();
    if( (screen = SDL_SetVideoMode( WIDTH, HEIGHT, vi->vfmt->BitsPerPixel, sdl_flags)) == NULL) {
        fprintf( stderr, "Can't set video mode: %s\n", SDL_GetError());
        exit( 1);
    }   

    SDL_WM_SetCaption( "Tennix " VERSION, "Tennix");
    SDL_ShowCursor( SDL_DISABLE);
    SDL_EnableKeyRepeat (SDL_DEFAULT_REPEAT_DELAY, 1);

    init_graphics();
    init_sound();
    init_joystick();

#ifdef ENABLE_FPS_LIMIT
    frames = 0;
    ft = SDL_GetTicks();
#endif

    if (benchmark) {
        GameState* g = gamestate_new();
        g->player1.type = PLAYER_TYPE_AI;
        g->player2.type = PLAYER_TYPE_AI;
        g->timelimit = BENCHMARK_TIMELIMIT*1000;
        gameloop(g);
        free(g);
        exit(0);
    }

    i = 0;
    /* Sliding initialization */
    ticks = SDL_GetTicks();
    slide = slide_start = get_image_width(GR_SIDEBAR);
    slide_direction = 0;
    while(!quit) {
        /* State transitions */
        switch (state) {
            case MENU_STATE_STARTED:
                state = MENU_STATE_SLIDE_TO_MAINMENU;
                break;
            case MENU_STATE_SLIDE_TO_MAINMENU:
                slide = slide_start;
                slide_direction = -1;
                state = MENU_STATE_SLIDE_TO_MAINMENU_IN_PROGRESS;
                break;
            case MENU_STATE_SLIDE_TO_MAINMENU_IN_PROGRESS:
                if (slide == 0) {
                    slide_direction = 0;
                    state = MENU_STATE_MAINMENU;
                }
                break;
            case MENU_STATE_MAINMENU:
                free(prepared_game);
                prepared_game = NULL;
                break;
            case MENU_STATE_SLIDE_TO_OPTIONS:
                slide = 1;
                slide_direction = 3;
                state = MENU_STATE_SLIDE_TO_OPTIONS_IN_PROGRESS;
                break;
            case MENU_STATE_SLIDE_TO_OPTIONS_IN_PROGRESS:
                if (slide == slide_start) {
                    start_fade();
                    state = MENU_STATE_OPTIONS;
                }
                break;
            case MENU_STATE_OPTIONS:
                /* Prepare a new game */
                if (prepared_game == NULL) {
                    prepared_game = gamestate_new();
                    btn_player1.text = "Keyboard (W-S-D)";
                    btn_player2.text = "Computer (AI)";
                }
                break;
            case MENU_STATE_SLIDE_TO_GAME:
                /*slide = 1;
                slide_direction = 2;
                state = MENU_STATE_SLIDE_TO_GAME_IN_PROGRESS;
                break;
            case MENU_STATE_SLIDE_TO_GAME_IN_PROGRESS:
                if (slide == slide_start) {
                    state = MENU_STATE_GAME;
                }*/
                state = MENU_STATE_GAME;
                break;
            case MENU_STATE_SLIDE_TO_RESUME:
                slide = 1;
                slide_direction = 2;
                state = MENU_STATE_SLIDE_TO_RESUME_IN_PROGRESS;
                break;
            case MENU_STATE_SLIDE_TO_RESUME_IN_PROGRESS:
                if (slide == slide_start) {
                    state = MENU_STATE_RESUME;
                }
                break;
            case MENU_STATE_GAME:
                if (prepared_game == NULL) {
                    fprintf(stderr, "Game not yet prepared!\n");
                    exit(EXIT_FAILURE);
                }
                /* Cancel a possibly started game */
                free(current_game);
                current_game = prepared_game;
                prepared_game = NULL;
                /* no break - we are continuing with "resume" */
            case MENU_STATE_RESUME:
                if (current_game == NULL) {
                    fprintf(stderr, "Cannot resume game!\n");
                    exit(EXIT_FAILURE);
                }
                start_fade();
                gameloop(current_game);
                SDL_Delay(150);
                while(SDL_PollEvent(&e));
#ifdef ENABLE_FPS_LIMIT
                frames = 0;
                ft = SDL_GetTicks();
#endif
                start_fade();
                state = MENU_STATE_SLIDE_TO_MAINMENU;
                break;
            case MENU_STATE_SLIDE_TO_QUIT:
                slide = 1;
                slide_direction = 3;
                state = MENU_STATE_SLIDE_TO_QUIT_IN_PROGRESS;
                break;
            case MENU_STATE_SLIDE_TO_QUIT_IN_PROGRESS:
                if (slide == slide_start) {
                    state = MENU_STATE_QUIT;
                }
                break;
            case MENU_STATE_QUIT:
                quit = true;
                break;
            default:
                fprintf(stderr, "State error: %d\n", state);
                exit(EXIT_FAILURE);
        }

        /* Sliding */
        if (SDL_GetTicks() > ticks + 20) {
            if (slide >= 1 && slide <= slide_start) {
                slide += slide_direction+(slide_direction*slide/(sqrt(2*slide)));
                slide = MAX(0, MIN(slide_start, slide));
            } else if (slide_direction != 0) {
                slide_direction = 0;
            }
            ticks = SDL_GetTicks();
        }

        /* Graphics */
#ifdef DEBUG
        if (state != MENU_STATE_OPTIONS) {
            fill_image_offset(GR_FOG, 0, 0, WIDTH, HEIGHT, -i, 0);
        }
#endif
        show_image(GR_SIDEBAR, WIDTH-get_image_width(GR_SIDEBAR)+slide, 0, 255);
        show_image(GR_TENNIXLOGO, WIDTH-get_image_width(GR_SIDEBAR)-10, 20-slide, 255);
        if (state != MENU_STATE_OPTIONS) {
            /* Main Menu */
            show_image(GR_BTN_PLAY, WIDTH-get_image_width(GR_BTN_PLAY)+slide+(slide/7)+3-(3*(btn_hovering==MENU_START)), 150, 255);
            if (current_game != NULL) {
                show_image(GR_BTN_RESUME, WIDTH-get_image_width(GR_BTN_RESUME)+slide+(slide/7)+3-(3*(btn_hovering==MENU_RESUME)), 230, 255);
                font_draw_string(GR_DKC2_FONT, "current match:", 10, 10, 0, 0);
                font_draw_string(GR_DKC2_FONT, current_game->sets_score_str, 10, 40, 0, 0);
            } else {
                font_draw_string(GR_DKC2_FONT, "Tennix " VERSION, 10, 10, 0, 0);
            }
            font_draw_string(GR_DKC2_FONT, URL, 10, HEIGHT-10-get_image_height(GR_DKC2_FONT), 0, 0);
            show_image(GR_BTN_QUIT, WIDTH-get_image_width(GR_BTN_QUIT)+slide+(slide/7)+3-(3*(btn_hovering==MENU_QUIT)), 350, 255);
        } else {
            /* Options screen */
            show_image(GR_TENNIXLOGO, WIDTH-get_image_width(GR_SIDEBAR)-10, 20, 255);
            draw_button_object(btn_back, mx, my);
            draw_button_object(btn_start, mx, my);
            if (prepared_game != NULL) {
                fill_image(prepared_game->court_type, MENU_OPTIONS_BORDER, MENU_OPTIONS_BORDER*2, get_image_width(GR_STADIUM), get_image_height(GR_STADIUM));
                show_image(GR_STADIUM, MENU_OPTIONS_BORDER, MENU_OPTIONS_BORDER*2, 255);
                draw_button_object(btn_court_change, mx, my);
                font_draw_string(GR_DKC2_FONT, "Location", MENU_OPTIONS_BORDER, MENU_OPTIONS_BORDER, 0, 0);
                draw_button_object(btn_player1, mx, my);
                draw_button_object(btn_player2, mx, my);
                font_draw_string(GR_DKC2_FONT, "Player 1", btn_player1.x, btn_player1.y-MENU_OPTIONS_BORDER, 0, 0);
                font_draw_string(GR_DKC2_FONT, "Player 2", btn_player2.x, btn_player2.y-MENU_OPTIONS_BORDER, 0, 0);
            }
        }

        SDL_PollEvent( &e);
        if( e.type == SDL_QUIT) {
            state = MENU_STATE_SLIDE_TO_QUIT;
            /*break;*/
        }

        keys = SDL_GetKeyState( NULL);
        mb = SDL_GetMouseState( &mx, &my);

        btn_hovering_old = btn_hovering;
        if (state == MENU_STATE_MAINMENU) {
            btn_hovering = M_POS_DECODE(mx, my);
            if (current_game == NULL) {
                btn_hovering &= ~MENU_RESUME;
            }
        } else if (state == MENU_STATE_OPTIONS) {
            if (M_POS_BUTTON(btn_back, mx, my)) {
                btn_hovering = MENU_QUIT;
            } else if (M_POS_BUTTON(btn_start, mx, my)) {
                btn_hovering = MENU_START;
            } else if (M_POS_BUTTON(btn_court_change, mx, my)) {
                btn_hovering = MENU_COURT_CHANGE;
            } else if (M_POS_BUTTON(btn_player1, mx, my)) {
                btn_hovering = MENU_PLAYER1;
            } else if (M_POS_BUTTON(btn_player2, mx, my)) {
                btn_hovering = MENU_PLAYER2;
            } else {
                btn_hovering = 0;
            }
        } else {
            /* No menu screen - no hovering. */
            btn_hovering = 0;
        }
#ifndef MAEMO /* On Maemo, we cannot really "hover" (touchscreen!) */
        if (btn_hovering_old != btn_hovering && btn_hovering != 0) {
#ifdef HAVE_VOICE_FILES
            if (btn_hovering == MENU_QUIT) {
                play_sample(VOICE_QUIT_IT);
            } else if (btn_hovering == MENU_START) {
                play_sample(VOICE_NEW_GAME);
            } else {
                play_sample(SOUND_MOUSEOVER);
            }
#else
            play_sample(SOUND_MOUSEOVER);
#endif
        }
#endif
   
        if( keys[SDLK_ESCAPE] || keys['q']) {
            /* FIXME: do the state thingie! */
            break;
        }

        if( keys['f']) {
            SDL_WM_ToggleFullScreen( screen);
        }
   
#ifndef MAEMO /* No mouse cursor on Maemo (we have a touchscreen) */
        if (state == MENU_STATE_MAINMENU || state == MENU_STATE_OPTIONS) {
            show_sprite( GR_RACKET, ((mb&SDL_BUTTON( SDL_BUTTON_LEFT))>0)+(((mb&SDL_BUTTON( SDL_BUTTON_RIGHT))>0)*2), 4, mx, my, 255);
        }
#endif

        /* Draw the "real" mouse coordinates */
        /*rectangle(mx-1, my-1, 2, 2, 255, 255, 255);*/

        /* Store the screen, because we are fading after this screen update */
        /*if (!(mb & SDL_BUTTON(SDL_BUTTON_LEFT)) && btn_hovering != MENU_NONE && mouse_pressed == true) store_screen();*/

        updatescr();

        if( mb & SDL_BUTTON(SDL_BUTTON_LEFT)) {
            mouse_pressed = true;
        } else if (mouse_pressed == true) {
            /* Mouse button released */
            if (state == MENU_STATE_MAINMENU || state == MENU_STATE_OPTIONS) {
#ifdef HAVE_VOICE_FILES
                if (btn_hovering == MENU_START) {
                    play_sample(VOICE_LETS_GO);
                } else {
                    play_sample(SOUND_MOUSECLICK);
                }
#else
                play_sample(SOUND_MOUSECLICK);
#endif
            }
            if (state == MENU_STATE_MAINMENU) {
                switch (btn_hovering) {
                    case MENU_START:
                        state = MENU_STATE_SLIDE_TO_OPTIONS;
                        break;
                    case MENU_RESUME:
                        state = MENU_STATE_SLIDE_TO_RESUME;
                        break;
                    case MENU_QUIT:
                        state = MENU_STATE_SLIDE_TO_QUIT;
                        break;
                }
            } else if (state == MENU_STATE_OPTIONS) {
                switch (btn_hovering) {
                    case MENU_START:
                        state = MENU_STATE_SLIDE_TO_GAME;
                        break;
                    case MENU_QUIT:
                        state = MENU_STATE_SLIDE_TO_MAINMENU;
                        break;
                    case MENU_COURT_CHANGE:
                        prepared_game->court_type++;
                        if (prepared_game->court_type > GR_CTT_LAST) {
                            prepared_game->court_type = GR_CTT_FIRST;
                        }
                        break;
                    case MENU_PLAYER1:
                        switch (prepared_game->player1.type) {
                            case PLAYER_TYPE_HUMAN:
                                prepared_game->player1.type = PLAYER_TYPE_AI;
                                btn_player1.text = "Computer (AI)";
                                break;
                            case PLAYER_TYPE_AI:
                                prepared_game->player1.type = PLAYER_TYPE_HUMAN;
                                btn_player1.text = "Keyboard (W-S-D)";
                                break;
                        }
                        break;
                    case MENU_PLAYER2:
                        switch (prepared_game->player2.type) {
                            case PLAYER_TYPE_HUMAN:
                                prepared_game->player2.type = PLAYER_TYPE_AI;
                                btn_player2.text = "Computer (AI)";
                                break;
                            case PLAYER_TYPE_AI:
                                prepared_game->player2.type = PLAYER_TYPE_HUMAN;
                                btn_player2.text = "Keyboard (O-L-K)";
                                break;
                        }
                        break;
                }
            }
            mouse_pressed = false;
        }
        i++;
#ifdef ENABLE_FPS_LIMIT
        while (frames*1000.0/((float)(SDL_GetTicks()-ft+1))>(float)(DEFAULT_FPS)) {
            SDL_Delay(10);
        }
        frames++;
#endif
    }

    uninit_graphics();
    uninit_joystick();

    SDL_Quit();
    return 0;
}

